### PR TITLE
Add wrapper for vtkImageDilateErode3D filter for UniformGrid Datasets

### PR DESCRIPTION
**Pull request by njneeteson**
### Overview

Wrapper for vtkImageDilateErode3D implemented for UniformGrid Datasets.

Resolves: #2179 

#### Filters

In pyvista/core/filters/uniform_grid.py:

Implemented a new method "image_dilate_erode" that wraps vtkImageDilateErode. Input parameters are the dilate and erode values and the diameter of the elliptical structural element in each dimension (they call it a kernel or a neighborhood in the VTK docs).

#### Tests

1. Test that it runs with and without specifying scalars.
2. Test that it correctly dilates a small volume of 0s with a single 1.
3. Test that it correctly erodes a 1 out of a volume of 0s with a single 1.

Test 1 is generic, tests 2 and 3 will capture changes in the behaviour of the underlying VTK class should that happen.

### CODE REVIEW

File: pyvista/core/filters/uniform_grid.py
Comment (tkoyama010) [inline pyvista/core/filters/uniform_grid.py:168]:
```suggestion
        elliptical footprint, and only erodes/dilates on the boundary of the
```

File: pyvista/core/filters/uniform_grid.py
Comment (tkoyama010) [inline pyvista/core/filters/uniform_grid.py:170]:
```suggestion
        It can degenerate to a 2 or 1-dimensional filter by setting the kernel
```

File: pyvista/core/filters/uniform_grid.py
Comment (tkoyama010) [inline pyvista/core/filters/uniform_grid.py:190]:
```suggestion
            When scalars are specified, this is the preferred array
```

File: pyvista/core/filters/uniform_grid.py
Comment (tkoyama010) [inline pyvista/core/filters/uniform_grid.py:176]:
```suggestion
            Dilate value in the dataset. Default: ``1``.
```

File: pyvista/core/filters/uniform_grid.py
Comment (tkoyama010) [inline pyvista/core/filters/uniform_grid.py:179]:
```suggestion
            Erode value in the dataset. Default: ``0``.
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
Thanks for adding this. LGTM with some small suggestions.

File: pyvista/core/filters/uniform_grid.py
Comment (adeak) [reply pyvista/core/filters/uniform_grid.py:176]: I'm not certain the suggestion is strictly better here. Same for `erode`.

Comment (adeak) [reply pyvista/core/filters/uniform_grid.py:176]: But it seems I'm wrong, sorry. ITK also uses "erode value" https://itk.org/ITKExamples/src/Filtering/BinaryMathematicalMorphology/ErodeABinaryImage/Documentation.html

**Comment by adeak** [comment]
I've pushed some minor style changes. @tkoyama010 I think CI is failing because there might be an `apt update` missing from the doc build action, at least it's looking for mesa packages that aren't there. Could you please take a look? I don't know CI stuff.

**Comment by njneeteson** [comment]
Thanks - I installed pre-commit and tried to install the hooks so hopefully these PEP8 formatting problems get caught before I commit/push in the future.

**Comment by adeak** [comment]
For some reason flake8 didn't catch any of that (perhaps because they were in tests?), and while isort picked up the import placement, it didn't trip CI for some reason.

In any case we'll soon have `black` which will make all these considerations moot :)

**Comment by njneeteson** [comment]
I did a real number on my local forked repo attempting to resolve the merge conflicts after the flake8/black PR merge into main (ended up just having to delete the whole thing and re-clone.. no idea what went wrong but git did not like the way I resolved the merge conflict and would not let me do anything), but thankfully the github web interface here let me salvage the branch :P

Pending the CI succeeding I think this should be ready for final review / merging.

In the future, will contributors need to run black locally on their code before submitting a PR, or how is that going to be incorporated? Will the contribution guide be updated to explain what we need to do pre-PR?

**Comment by tkoyama010** [comment]
@njneeteson Thank you for resolving a very difficult conflict. That is exactly what we are discussing in https://github.com/pyvista/pyvista/pull/2180 now.
I would be grateful for you to join the discussion.

**Comment by njneeteson** [comment]
No problem.. resolving the conflict in this case was pretty easy in the web GUI. Delete the bad-style code that `black` wanted to replace and reformat my own code to be compliant. It helped that there aren't that many methods in `uniform_grid.py` and mostly they ahve few enough parameters that the def lines don't go over 100 chars. I imagine this will be a total nightmare for people who edited larger files.

**Comment by njneeteson** [comment]
OK with the #2212 fix I have merged main back into this feature branch, ran `make format` to fix the style, and pushed for the last time. It should be all good to go assuming the CI passes.

File: pyvista/core/filters/uniform_grid.py
Comment (adeak) [inline pyvista/core/filters/uniform_grid.py:245]: Slightly more pythonic with the added benefit of crashing if `kernel_size` has more than 3 elements:
```suggestion
        alg.SetKernelSize(*kernel_size)
```

File: pyvista/core/filters/uniform_grid.py
Comment (adeak) [inline pyvista/core/filters/uniform_grid.py:212]: Is this "thresholded" a typo from copy-pasting? Seems off.

**Review by adeak — COMMENTED** [review]
Thanks for persevering in the `black` chaos. Great work on the PR!

One potential issue and one mild code style suggestion. If we resolve the docstring issue this is ready to merge.

File: pyvista/core/filters/uniform_grid.py
Comment (njneeteson) [reply pyvista/core/filters/uniform_grid.py:212]: [Thresholded](https://en.wiktionary.org/wiki/thresholded) is a word but I can change it if you think it's weird/awkward.

File: pyvista/core/filters/uniform_grid.py
Comment (njneeteson) [reply pyvista/core/filters/uniform_grid.py]: I agree, that's better.

**Comment by njneeteson** [comment]
I've switched this to draft in light of #2217 - I assume this filter is affected as well and the unit tests do not currently cover attempting to dilate/erode using `cell_data`

File: pyvista/core/filters/uniform_grid.py
Comment (adeak) [reply pyvista/core/filters/uniform_grid.py:212]: Sorry, I phrased that ambiguously. I meant that the filter doesn't threshold, it erodes/dilates, right? If instead you mean that scalars get transformed in nontrivial ways this should be explained somewhere in the docstring.

Comment (njneeteson) [reply pyvista/core/filters/uniform_grid.py:212]: Oh you are right, this is a copy-paste error. I will fix it in the next commit. Thanks!

**Comment by akaszynski** [comment]
I marked this ready for review based on the comments in #2217. It appears that VTK does not support cell data within its image filters, and we’ll need a workaround for it (or at least document the behavior) in a follow up PR.

**Review by akaszynski — APPROVED** [review]
Excellent implementation. I’d like to have this merged soon to have it out for our next release. I know that the cell_data issue remains, but for the time being (and as it affects multiple filters), let’s hold off until a later PR.

**Comment by akaszynski** [comment]
Merging as I'm happy with the implementation and there has been no follow-up comments for over a week.

**Pull request merged**
